safenode/src/domain/storage/disk_backed_record_store.rs

```diff
@@ -0,0 +1,483 @@
+                let record = Record {
+                    key: k.clone(),
+                    value: contents,
+                    publisher: None,
+                    expires: None,
+                };
+                Some(Cow::Owned(record))
+            }
+            Err(err) => {
+                error!("Error while reading file. filename: {filename}, error: {err:?}");
+                None
+            }
+        }
+    }
+
+    fn put(&mut self, r: Record) -> Result<()> {
+        trace!("PUT request for Record key: {:?}", r.key);
+        if r.value.len() >= self.config.max_value_bytes {
+            warn!(
+                "Record not stored. Value too large: {} bytes",
+                r.value.len()
+            );
+            return Err(Error::ValueTooLarge);
+        }
+
+        // todo: the key is not tied to the value it contains, hence the value can be overwritten
+        // (incase of dbc double spends etc), hence need to deal with those.
+        // Maybe implement a RecordHeader to store the type of data we're storing?
+        if self.disk_backed_records.contains(&r.key) {
+            debug!(
+                "Record with key {:?} already exists, not overwriting.",
+                r.key
+            );
+            return Ok(());
+        }
+
+        let num_records = self.disk_backed_records.len();
+        if num_records >= self.config.max_records {
+            warn!("Record not stored. Maximum number of records reached. Current num_records: {num_records}");
+            return Err(Error::MaxRecords);
+        }
+
+        let filename = Self::key_to_hex(&r.key);
+        let file_path = self.config.storage_dir.join(&filename);
+        match fs::write(file_path, r.value) {
+            Ok(_) => {
+                trace!("Wrote record to disk! filename: {filename}");
+                let _ = !self.disk_backed_records.insert(r.key);
+                Ok(())
+            }
+            Err(err) => {
+                error!("Error writing file. filename: {filename}, error: {err:?}");
+                Ok(())
+            }
+        }
+    }
+
+    fn remove(&mut self, k: &Key) {
+        let filename = Self::key_to_hex(k);
+        let file_path = self.config.storage_dir.join(&filename);
+        match fs::remove_file(file_path) {
+            Ok(_) => {
+                trace!("Removed record from disk! filename: {filename}");
+                let _ = self.disk_backed_records.remove(k);
+            }
+            Err(err) => {
+                error!("Error while removing file. filename: {filename}, error: {err:?}");
+            }
+        }
+    }
+
+    fn records(&self) -> Self::RecordsIter<'_> {
+        let mut records = Vec::new();
+        for key in self.disk_backed_records.iter() {
+            if let Some(record) = self.get(key) {
+                records.push(record);
+            }
+        }
+        records.into_iter()
+    }
+
+    fn add_provider(&mut self, record: ProviderRecord) -> Result<()> {
```

Comment:
this shall only get called when `PoviderRecord` is enabled ?
Which won't be used in the current cfg?
Maybe just leave it as `un-implemented` ?

Same to other ProviderRecord realted impls

Reply:
Currently debugging an issue that raised after marking the methods as todo!()

Reply:
The `kad_cfg.set_provider_publication_interval()` caused the node to trigger the provider APIs. Removing the cfg fixes the issue.